octoprint_mrbeam/static/js/helpers/working_area_helper.js

```diff
@@ -308,29 +309,49 @@ class WorkingAreaHelper {
                 }`;
             } else {
                 const v = event.target.value;
-                const idxDelimiter = v.search(new RegExp(options.delimiter));
-                if (selStart <= idxDelimiter) {
-                    const v1 = v.substring(0, idxDelimiter);
-                    const newV1 = parseFloat(v1) + val;
-                    event.target.value = `${newV1.toFixed(
-                        options.digits
-                    )}${v.substring(idxDelimiter)}`;
-                } else {
-                    const d = idxDelimiter + options.delimiter.length;
-                    const v2 = v.substring(d);
-                    const newV2 = parseFloat(v2) + val;
-                    event.target.value = `${v.substring(0, d)}${newV2.toFixed(
-                        options.digits
-                    )}`;
+
+                let parts = this._getSubstringAtIndex(
+                    v,
+                    selStart,
+                    options.delimiter
+                );
+                const newV1 = (parseFloat(parts[1]) + val).toFixed(
+                    options.digits
+                );
+                event.target.value = `${parts[0]}${newV1}${parts[2]}`;
+                // maintain cursor position
+                if (selStart > parts[0].length) {
+                    cursorShift = newV1.length - parts[1].length;
                 }
             }
             // restore caret position
-            event.target.selectionStart = selStart;
-            event.target.selectionEnd = selEnd;
+            event.target.selectionStart = selStart + cursorShift;
+            event.target.selectionEnd = selEnd + cursorShift;
             return false; // swallow the default action
         }
         return true;
     }
+
+    static _getSubstringAtIndex(input, index, delimiter = "[^a-zA-Z0-9.,]+") {
+        const iter = input.matchAll(delimiter);
+        let idx1 = 0;
+        let idx2 = input.length;
+        let del;
+        while ((del = iter.next()) !== null) {
+            if (del.done) break;
+            if (del.value.index < index) {
+                idx1 = del.value.index + del.value[0].length;
+            } else {
+                idx2 = del.value.index;
+                break;
+            }
+        }
+        const strBefore = input.substring(0, idx1);
+        const v1 = input.substring(idx1, idx2);
+        const strAfter = input.substring(idx2);
+        //console.debug("idx, idx1 -> idx2", index, idx1, idx2);
+        return [strBefore, v1, strAfter];
+    }
```

Comment:
That also changes any weird separator to the default separator

Reply:
Found a nice reduce one-liner for `new_string`
```js
let delimiter = ','
[1, 2, 3].concat(5, [6, 7]).reduce((x, y) => x + delimiter + y)
// "1,2,3,5,6,7"
```

Reply:
One should probably also filter out the empty values after the split (in case of 2 adjacent separators) :
```js
before_split = before_cursor.split(delimiter).filter(e=>e)
```

Reply:
The main issue with replacing the separator is that the cursor might jump from one value to the next after incrementing

Reply:
What's the benefit ?

Reply:
> That also changes any weird separator to the default separator

? what is a weird separator ?

Reply:
> Found a nice reduce one-liner for `new_string`
> 
> ```js
> let default_delimiter = ','
> [1, 2, 3].concat(5, [6, 7]).reduce((x, y) => x + default_delimiter + y)
> // "1,2,3,5,6,7"
> ```

does this really do what _getSubstringAtIndex() is doing?  I don't understand...

Reply:
The benefit is code readability, super important!

By weird separator I mean anything that isn't what we expect like a `////` as the default `delimitor` variable would suggest.

It's not meant to just replace this specific function but also part of the code where the function is referenced.

What it does is
- parsing the line with the given separators
- selecting the item you have your cursor on
- increment it with some value
- join together the previously parsed items with a default separator (can be different than what the user typed in)